internal/dataplane/fallback/fallback.go

```diff
@@ -102,32 +115,42 @@ func (g *Generator) GenerateBackfillingBrokenObjects(
 
 	if lastValidCacheSnapshot == nil {
 		g.logger.V(util.DebugLevel).Info("No previous valid cache snapshot found, skipping backfilling")
-		return fallbackCache, nil
+		return fallbackCache, nil, nil
 	}
 
 	// Build a graph from the last valid cache snapshot.
-	lastValidGraph, err := g.cacheGraphProvider.CacheToGraph(*lastValidCacheSnapshot)
+	lastValidGraph, diag, err := g.cacheGraphProvider.CacheToGraph(*lastValidCacheSnapshot)
 	if err != nil {
-		return store.CacheStores{}, fmt.Errorf("failed to build cache graph: %w", err)
+		return store.CacheStores{}, nil, fmt.Errorf("failed to build cache graph: %w", err)
 	}
 
 	// Backfill the affected objects from the last valid cache snapshot.
 	for _, affectedObject := range affectedObjects {
 		objectsToBackfill, err := lastValidGraph.SubgraphObjects(affectedObject)
 		if err != nil {
-			return store.CacheStores{}, fmt.Errorf("failed to find dependants for %s: %w", affectedObject, err)
+			return store.CacheStores{}, nil, fmt.Errorf("failed to find dependants for %s: %w", affectedObject, err)
 		}
 
-		for _, obj := range objectsToBackfill {
+		backdiag = make([]diagnostics.FallbackDiagnostic, len(objectsToBackfill))
+		for i, obj := range objectsToBackfill {
 			if err := fallbackCache.Add(obj); err != nil {
-				return store.CacheStores{}, fmt.Errorf("failed to add %s to the cache: %w", GetObjectHash(obj), err)
+				return store.CacheStores{}, diag, fmt.Errorf("failed to add %s to the cache: %w", GetObjectHash(obj), err)
 			}
 			g.logger.V(util.DebugLevel).Info("Backfilled object to fallback cache from previous valid cache snapshot",
 				"object_kind", obj.GetObjectKind(),
 				"object_name", obj.GetName(),
 				"object_namespace", obj.GetNamespace(),
 			)
+			gvk := obj.GetObjectKind().GroupVersionKind()
+			backdiag[i] = diagnostics.FallbackDiagnostic{
+				GroupKind:     fmt.Sprintf("%s/%s", gvk.Group, gvk.Kind),
+				Namespace:     obj.GetNamespace(),
+				Name:          obj.GetName(),
+				ID:            string(obj.GetUID()),
+				Status:        "backfilled",
+				CausingObject: affectedObject.String(),
+			}
```

Comment:
1. This seems to incorrectly use the `backdiag` by overwriting it for each `affectedObject`, no? Line 134 allocates `len(objectsToBackfill)`, then the inner for loop fills those in, and the cycle repeats for next `affectedObject`.
2. Refactor idea: this block could be excracted into a function taking e.g. `affectedObjects` and `lastValidGraph` as args.